### PR TITLE
[No QA] Move version-bump and automerge code to its own workflow

### PR DESCRIPTION
**Pull request by roryabraham**
### Details
This pull request does a few things:

- At an abstract level, it establishes a pattern by which we can make a "utility workflow" or "subroutine", which can be synchronously or asynchronously executed by another workflow. Nifty! Here's how it works:

    - Make the utility workflow execute on the [workflow_dispatch event](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_dispatch). That means that the workflow can be manually triggered via an arbitrary workflow_dispatch event.
    - Use https://github.com/convictional/trigger-workflow-and-wait to trigger the specified workflow with the provided inputs and wait until that workflow completes. Alternatively, you can choose not to wait (we don't do that anywhere in this PR).
        - GitHub Actions are designed such that, by default, events triggered by one workflow cannot cause another to execute. This is by design to prevent circular workflows (infinite loops). However, we can work around this by using the `OS_BOTIFY_TOKEN` instead of the default `GITHUB_TOKEN`.
        
Ultimately, this should promote code reuse and greatly simplify the handling of race conditions, the creation of new workflows, and the update of branches.

Now, specifically what this PR does:

##### Creates a new workflow called `createNewVersion` that:

1. Creates a new version with a specified `SEMVER_LEVEL`
2. Creates a version-bump pull request against `main` to update the app version.
3. Approves and automatically merges that pull request.

It's important to note that the `createNewVersion` workflow uses softprops/turnstyle to ensure that only one instance of the workflow is executing at a time. By doing so, it ensures that one version is created and committed to `main` before we allow the next version to be created.

##### Updates the `bumpVersion` action to use the local version instead of tags

The `bumpVersion` action previously used tags to prevent version conflicts due to race conditions. i.e: when a version was created we would immediately tag it and push the tags. Then when creating a new version we would fetch the tags, get the highest one, and do that + 1. The problem created by that method was that we could not "register" a new version without pushing tags, and pushing tags triggers a staging deploy.

Now, we're using a "subroutine" and softprops/turnstyle to prevent race conditions, and we don't want to deploy to staging until the staging branch has been successfully updated. So I changed this action to once again just use the local version +1.

##### Replaces the `automerge` workflow with a new `updateProtectedBranch` subroutine

- When we want to update a protected branch, we have to create an automerge a pull request to update that branch. This workflow does that synchronously.
- `staging` can only be updated from `main`, and `production` can only be updated from `staging`
- This workflow can be synchronously executed from within our `createNewVersion` workflow, to prevent version conflict race conditions.
- This workflow also means we no longer need the `automerge` workflow.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/157916 https://github.com/Expensify/Expensify/issues/163576

### Testing plan

1. Make sure the `StagingDeployCash` is _not_ locked
2. Create a dummy PR.
3. Merge this PR and the dummy PR in quick succession.
4. Both PRs should create a new `BUILD` version. There should be no version conflict. A tag for each version should _not_ be pushed yet.
5. For each, a bump-version PR should update the `main` branch, followed by another PR to update the `staging` branch from `main`.
6. Once the `staging` branch is updated, the `deploy` workflow should begin, create a tag for the updated `BUILD` version on the `staging` branch, and push tags.
7. Pushing tags should start the staging `platformDeploy` workflow. When that workflow completes, a comment should appear on this and the dummy PR stating that it was deployed to staging.
8. Both PRs should be added to the `StagingDeployCash`.
9. Lock the `StagingDeployCash`, and the `lockCashDeploys` workflow should begin.
10. A `PATCH` version should be created.
11. A PR to update the `main` branch should be created and automerged, followed by a PR to update the `staging` branch.
12. Once the PR to update `staging` is merged, the `deploy` workflow should begin, create a tag with the updated `PATCH` version, and push tags.
13. Pushing tags should start the staging `platformDeploy` workflow.
14. Create and merge another dummy PR. It should _not_ result in the creation of a new version, and should _not_ result in a deploy. We _should_ see a comment like: `This PR was not deployed to staging...`.
15. Write down all the un-QA'd PRs and unresolved deploy blockers on the open `StagingDeployCash`.
16. Check off all the items on the `StagingDeployCash` checklist, and close it with the `:shipit:` comment.
17. The `finishReleaseCycle` workflow should begin, and a pull request with `production` as the base and `staging` as the head should be created and automerged.
18. Once the PR to update `production` is merged, the `deploy` workflow should start, and create a new GH release for the `PATCH` version that was created in step 10.
19. Once that release is created, the `platformDeploy` workflow should begin a prod deploy. Once that completes, a comment should appear on this PR, the dummy PR, and all the PRs that were on the now-closed `StagingDeployCash` stating that they were deployed to production.
20. The `finishReleaseCycle` workflow should also bump the `BUILD` version and create a PR to update `staging` from main. This should result in a staging deploy.
21. The `finishReleaseCycle` workflow should also create a new `StagingDeployCash` issue, with the `BUILD` version from the previous step.
22. The dummy PR from step 14 should be included in the PR checklist for the new `StagingDeployCash`.
23. As a cleanup step, manually add any unverified PRs to the new `StagingDeployCash` (you wrote these down in step 15). The deploy blockers will no longer be deploy blockers, but should still be worked on with as much urgency as possible, as they will now be bugs on production.
 
#### Merging regular PR (w/ race condition)
- [ ] Staging Deploy _not_ locked
- Two PRs:
    1.
    2.
- `preDeploy` workflow run:
    - [ ] Created BUILD version
    - [ ] Updated main:
        - PR:
        - `updateProtectedBranch` workflow run:
    - [ ] Updated staging:
        - PR:
        - `updateProtectedBranch` workflow run:
- `deploy` workflow run:
    - [ ] Tag pushed?
- `platformDeploy` workflow run:
- [ ] `StagingDeployCash` updated with new PRs and `BUILD` version?
- [ ] "Deployed to staging" comment on merged PRs?
 
#### Locking `StagingDeployCash`
- `lockCashDeploys` workflow run:
    - [ ] Patch version created?
    - [ ] Updated main:
        - PR:
        - `updateProtectedBranch` workflow run:
    - [ ] Updated staging:
        - PR:
        - `updateProtectedBranch` workflow run:
    - [ ] `StagingDeployCash` updated with new `PATCH` version?

#### Merging while `StagingDeployCash` is locked
- Dummy PR:
- `preDeploy` workflow:
    - [ ] "Not deployed to staging" comment?
    - [ ] StagingDeployCash not updated?
    
#### Prod deploy
- [ ] `finishReleaseCycle` workflow run:
- [ ] Updated production:
    - PR:
    - `updateProtectedBranch` workflow run:
- [ ] `deploy` workflow run (prod):
- [ ] New release created w/ version matching closed `StagingDeployCash`
- [ ] `platformDeploy` workflow run (prod):
- [ ] "Deployed to production" comment on PRs from `StagingDeployCash` checklist?
- [ ] `BUILD` version created?
- [ ] `main` updated with new `BUILD` version?
    - PR:
    - `updateProtectedBranch` workflow run:
- [ ] `Staging` updated with new `BUILD` version?
    - PR:
    - `updateProtectedBranch` workflow run:
- [ ] `deploy` workflow run (staging):
- [ ] `platformDeploy` workflow run (staging):
    - [ ] Does the PR that was merged when `StagingDeployCash` was locked now say "deployed to staging"?
- [ ] New `StagingDeployCash` created:
    - [ ] new `BUILD` version?
    - [ ] Includes PR that was merged while `StagingDeployCash` was locked?

### CODE REVIEW

**Comment by roryabraham** [comment]
Okay, there are still more simplifications we can do after this, but we can move in increments. I _think_ this PR is enough to resolve the problem at hand (i.e: only deploy to staging when the staging branch is updated, and same for production). There are several more simplifications we can do after this before adding complexity with the CP flows.

**Comment by roryabraham** [comment]
I took this off draft, and I could use some reviews at this point. That said, the likelihood that this won't break anything and will just work first try is slim. To minimize disruptions, @AndrewGable and I were thinking we would schedule a day to live-test this, and let Applause skip QA that day. So please review but don't merge 🙂

File: .github/workflows/automerge.yml
Comment (AndrewGable) [inline .github/workflows/automerge.yml:33]: Oh nice PR - Is there anything we can do to get this merged or debug the failures?

File: .github/workflows/createNewVersion.yml
Comment (AndrewGable) [inline .github/workflows/createNewVersion.yml:91]: Bummer - Are there any alternatives that do? How hard would it be to write our own in JS?

**Review by AndrewGable — COMMENTED** [review]
Looking good! I like where this is headed for sure. It seems like there are still some TODOs left before we can talk about testing this, is that correct?

File: .github/workflows/automerge.yml
Comment (roryabraham) [reply .github/workflows/automerge.yml:33]: I can try to look into it some more, but kind of got stuck since I'm really unable to reproduce the failures that I'm seeing. Maybe a fresh set of eyes could help?

**Comment by roryabraham** [comment]
> there are still some TODOs left before we can talk about testing this, is that correct?

Correct, we need to either get those 3rd-party PRs merged or point at my fork of the repos for now. We could probably benefit from doing some more atomic testing of the updated 3rd-party actions using `Public-Test-Repo` before diving in to try and test this whole PR

File: .github/workflows/automerge.yml
Comment (AndrewGable) [reply .github/workflows/automerge.yml:33]: I can take a look but I am curious why we want to move this away from the label trigger? Is there a race condition? Or we just want to be able to wait on it?

File: .github/workflows/createNewVersion.yml
Comment (roryabraham) [reply .github/workflows/createNewVersion.yml:91]: Funny enough, there is a quirk with this in that the `GITHUB_TOKEN` cannot be used to trigger new actions (by design, to prevent infinite loops creating workflows). Before I realized that the workaround is to simply use a PAT (`OS_BOTIFY_TOKEN` in our case), I sort of wrote it off as a problem with the `trigger-workflow-and-wait` action. So I started writing my own, and have a [draft PR](https://github.com/Expensify/Expensify.cash/pull/2495). You can take a look ... it's a bit complex.

File: .github/workflows/automerge.yml
Comment (roryabraham) [reply .github/workflows/automerge.yml:33]: > Is there a race condition?

Yes, prettymuch. We want to eliminate the need to push tags or have a remote store of versions to resolve conflicts, and this PR is moving us to a world where the we can eliminate race conditions at the workflow level, and the JS can just bump the version using what it has locally. In order that that to work, we need to synchronously wait, not only for a new version to be created, but also for it to be merged. That way, when turnstyle allows the next new version to be created, we can again just use the local version on main and it will have been updated.

In order to do that, we need to treat `automerge.yml` as a "subroutine", just like we are for `createNewVersion.yml`

Comment (AndrewGable) [reply .github/workflows/automerge.yml:33]: Ok that makes sense, thank you for the explanation 👍

File: .github/workflows/createNewVersion.yml
Comment (AndrewGable) [reply .github/workflows/createNewVersion.yml:91]: So.. Should we use yours since it can work on mac-os and will simplify this workflow code at least?

Comment (Jag96) [reply .github/workflows/createNewVersion.yml:91]: +1, I think we should look into using the draft PR implementation for this then

File: .github/workflows/finishReleaseCycle.yml
Comment (Jag96) [inline .github/workflows/finishReleaseCycle.yml:113]: I'm not sure what this means, is this still a TODO?

**Review by Jag96 — COMMENTED** [review]
> We could probably benefit from doing some more atomic testing of the updated 3rd-party actions using Public-Test-Repo before diving in to try and test this whole PR

This sounds like a good idea, what specific tests need to be run?

File: .github/actions/bumpVersion/bumpVersion.js
Comment (deetergp) [inline .github/actions/bumpVersion/bumpVersion.js:59]: You can just do `"Update version to %s"` here. ([ref](https://docs.npmjs.com/cli/v7/commands/npm-version))

File: .github/workflows/createNewVersion.yml
Comment (deetergp) [inline .github/workflows/createNewVersion.yml:33]: NAB: Do you know if these are chained like `&&` in bash, in that the next command won't run unless the previous command exits with a return code of `0`?

**Review by deetergp — DISMISSED** [review]

File: .github/workflows/createNewVersion.yml
Comment (roryabraham) [reply .github/workflows/createNewVersion.yml:33]: That's a good question ... I do not know the answer.

File: .github/workflows/createNewVersion.yml
Comment (roryabraham) [reply .github/workflows/createNewVersion.yml:91]: Okay, in that case I will need to polish and test that draft PR. I believe it can be tested in `Public-Test-Repo` to validate that it works.

**Comment by roryabraham** [comment]
Got some movement on [this PR](https://github.com/pascalgn/automerge-action/pull/149) and updated it after the maintainer's review.

File: .github/workflows/automerge.yml
Comment (roryabraham) [inline .github/workflows/automerge.yml:31]: Honestly, I'm not sure if just pointing at a forked repo like this will work, or if I need to checkout that repo in the workflow or publish the action in that repo or something

Comment (roryabraham) [reply .github/workflows/automerge.yml]: Seems like this works just fine as-is: https://github.com/Andrew-Test-Org/Public-Test-Repo/pull/49/checks?check_run_id=2568980822

**Comment by roryabraham** [comment]
Okay, taking this off hold. Let's do this thang.

**Review by Jag96 — APPROVED** [review]

**Comment by Jag96** [comment]
@deetergp dismissing the requested changes here since they were for minor refactors and we're going to be testing this today. Will follow up w/ any minor fixes.

**Pull request merged**
